runtime/near-vm-runner/RUNTIMES.md

```diff
@@ -0,0 +1,214 @@
+linking). Only then it is possible to start executing the contract code.
+
+When executing a tiny function part of a larger contract these operations will dominate and
+contribute greatly to the observed latency of the contract execution. These overheads contribute to
+the fees paid by anybody using the protocol, making any unnecessary overhead a potential roadblock
+in NEAR protocol's adoption.
+
+### Throughput
+
+The faster we can execute the contract code, the more elaborate are the ideas that developers can
+express via contracts running on the NEAR protocol. The cheaper it is to execute contracts, the
+more approachable and usable the protocol is, so runtime performance is a direct consequence of the
+usability of the whole system as a whole.
+
+One of more costly operations that occurs in a typical contract quite often is accounting for gas
+costs, so it is quite important that the runtime gives sufficient flexibility to the NEAR
+validator to implement such operations efficiently. And, of course, general code quality also
+matters.
+
+# Evaluation
+
+We have evaluated a number of different implementations over time. Currently Wasmer with the
+universal engine and the single-pass backend is the default.
+
+## `wasmer-singlepass` codegen
+
+The `wasmer-singlepass` code generator is an implementation of a straightforward, direct
+translation from Wasm statements to machine code. This approach is linear over the input size by
+construction satisfying the requirements for preparation performance.
+
+A single-pass codegen implementation in general incurs a trade-off in runtime performance.
+WebAssembly is modelling a stack machine and an obviously correct implementation of such a machine
+would be to directly translate all the stack operations to native code as well. Targets running
+this code, on the other hand, are most likely an implementation of register machine where stack
+operations are not necessarily all that efficient.
+
+In order to improve the quality of the generated machine code, `wasmer-singlepass` maintains some
+state between the individual Wasm instructions. This state allows `wasmer-singlepass` to avoid some
+stack operations, pick more efficient instructions and sometimes avoid some operations such as NaN
+canonicalization altogether.
+
+Presence of this state comes at a cost of testability of implementation. The behaviour of the
+codegen can depend on prior Wasm instructions. Because of this there's a combinatorial explosion of
+the test cases that need to be considered and verified before confidence in the results can be
+established. Relatedly, at times it may also be difficult to validate changes made to the
+implementation of the `wasmer-singlepass` codegen. The global state is definitely a source of
+potential spooky action at a distance problems where changing code generation of a specific
+instruction affects correctness and behaviour of another one.
```

Comment:
I have trouble parsing this sentence. Is there a typo somewhere?

Reply:
Gave a shot at rewording this.